File: ios/Plugin/OryAuthPlugin.swift
import Foundation
import Capacitor


/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(OryAuthPlugin)
public class OryAuthPlugin: CAPPlugin {

    private var auth: OryAuth? = nil

    @objc func initialize(_ call: CAPPluginCall) {
        let basePath = call.getString("basePath") ?? ""
        OpenAPIClientAPI.basePath = basePath
        auth = OryAuth()
        call.resolve()
    }
    
    @objc func session(_ call: CAPPluginCall) {
        
        guard auth == nil else {
            call.reject("Can not get session. Initialize plugin before use.");
            return
        }
        
        auth!.session(completion: { active, error in
            call.resolve([
                "active": active
            ])
        })
    }
    
    @objc func signIn(_ call: CAPPluginCall) {
        
        guard auth != nil else {
            call.reject("Can not sign in. Initialize plugin before use.");
            return
        }
        
        do {
            let password = call.getString("password")
            let traits = call.getObject("traits")
            
            guard password != nil && traits != nil && traits?.index(forKey: "email") != nil else {
                call.reject("Specify a password and traits.email")
                return
            }
        
            let theJSONData = try JSONSerialization.data(withJSONObject: traits as Any, options: [])
            let jsonTraits = String(data: theJSONData, encoding: .utf8)!
        
            auth!.signin(password: password!, traits: jsonTraits) { active, error in
                guard error == nil else {
                    call.reject(error!.localizedDescription)
                    return
                }
                call.resolve([
                    "active": active ?? false
                ])
            }
        
        } catch {
            call.reject(error.localizedDescription)
        }

    }
    @objc func verify(_ call: CAPPluginCall) {
        
        guard auth != nil else {
            call.reject("Can not verify. Initialize plugin before use.");
            return
        }
        
        let id = call.getString("id");
        let code = call.getString("code");
        let email = call.getString("email");
        
        guard id == nil && code == nil else {
            call.reject("Specify a id and code")
            return
        }
        
        auth!.verify(id: id!, code: code!, email: email, completion: { state, error in
            guard error == nil else {
                call.reject(error!.localizedDescription)
                return
            }
            call.resolve(["state": state])
        })
    }
    
    @objc func logIn(_ call: CAPPluginCall) {
        
        guard auth != nil else {
            call.reject("Can not log in. Initialize plugin before use.");
            return
        }
    
        let identifier = call.getString("identifier") ?? ""
        let password = call.getString("password") ?? ""
        auth!.login(identifier: identifier, password: password, completion: { error in
            guard error == nil else {
                call.reject(error!.localizedDescription)
                return
            }
            call.resolve()
        })
    }
    
    @objc func logOut(_ call: CAPPluginCall) {
        
        guard auth != nil else {
            call.reject("Can not log out. Initialize plugin before use.");
            return
        }
        
        auth!.logout(completion: { error in
            guard error == nil else {
                call.reject(error!.localizedDescription)
                return
            }
            call.resolve()
        })
    }
}

File: android/src/main/java/es/frnd/capacitor/oryauth/OryAuthPlugin.java
package es.frnd.capacitor.oryauth;

import android.webkit.JavascriptInterface;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

import sh.ory.ApiClient;
import sh.ory.ApiException;
import sh.ory.Configuration;

@CapacitorPlugin(name = "OryAuth")
public class OryAuthPlugin extends Plugin {

    private OryAuth auth;

    private SessionTokenStore sessionTokenStore;

    @Override
    public void load() {
        sessionTokenStore = new SessionTokenStore(getContext());
        this.bridge.getWebView().addJavascriptInterface(this, "CapacitorOryAuthAndroidInterface");
        super.load();
    }

    @JavascriptInterface
    public String sessionToken() {
        return sessionTokenStore.get();
    }

    @PluginMethod()
    public void initialize(PluginCall call) {
        ApiClient defaultClient = Configuration.getDefaultApiClient();
        String basePath = call.getString("basePath");
        defaultClient.setBasePath(basePath);
        auth = new OryAuth(sessionTokenStore, defaultClient);
        call.resolve();
    }

    @PluginMethod()
    public void session(PluginCall call) {

        if (auth == null) call.reject("Initialize plugin before use.");

        JSObject ret = new JSObject();
        ret.put("active", auth.session());
        call.resolve(ret);
    }

    @PluginMethod()
    public void signIn(PluginCall call) {

        if (auth == null) call.reject("Initialize plugin before use.");

        String password = call.getString("password");
        JSObject traits = call.getObject("traits");

        try {
            String flowId = auth.signIn(password, traits.toString());
            JSObject res = new JSObject();
            res.put("verificationFlowId", flowId);
            call.resolve(res);
        } catch (ApiException e) {
            call.reject(e.getMessage());
        }

    }

    @PluginMethod()
    public void verify(PluginCall call) {

        if (auth == null) call.reject("Initialize plugin before use.");

        String id = call.getString("id");
        String code = call.getString("code");
        String email = call.getString("email");

        try {
            Object state = auth.verify(id, code, email);
            JSObject res = new JSObject();
            res.put("state", state);
            call.resolve(res);
        } catch (ApiException e) {
            call.reject(e.getMessage());
        }
    }

    @PluginMethod()
    public void logIn(PluginCall call) {

        if (auth == null) call.reject("Initialize plugin before use.");

        String identifier = call.getString("identifier");
        String password = call.getString("password");

        try {
            auth.login(identifier, password);
            call.resolve();
        } catch (ApiException e) {
            call.reject(e.getMessage());
        }

        System.out.println("logIn");
        call.resolve();
    }

    @PluginMethod()
    public void logOut(PluginCall call) {

        if (auth == null) call.reject("Initialize plugin before use.");

        try {
            JSObject res = new JSObject();
            res.put("state", auth.logout());
            call.resolve(res);
        } catch (ApiException e) {
            call.reject(e.getMessage());
        }
    }
}
